### PR TITLE
Overhaul configuration system

### PR DESCRIPTION
**Pull request by tskisner**
The current configuration system suffers from some challenges, including fragile / special-case parsing of container traits.  This work aims to make this area of the code robust and more future-proof:

* Move config tools into a submodule, extract some things from traits.py.

* Support arbitrary nested containers, including parsing of those and None values from the commandline.  Expand unit tests to included nested containers.

* Dedicated set of routines for translating back and forth from trait values and strings.

* Separate file-format code more cleanly from code that just works with the internal config dictionary format.

* Add support for YAML config files, along with existing JSON and TOML formats.

* Expand the `toast_config_verify` script to support loading and migrating old config files to new ones (or for converting between supported formats).

* Keep support for loading old config files, while printing warnings telling the user to update their configs.

As part of testing these changes, I checked (with `toast_config_verify`) that existing config files from CMB-S4 DC0 and from S.O. integration tests are loadable (with deprecation warnings).

### CODE REVIEW

File: src/toast/config/yaml.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    if isinstance(elem, list):
+        return (trait_to_string(elem), "list")
+
+    # This is a string, which might represent a quantity or
+    # some complicated container.  Convert to a value that
+    # we can use to determine the type.
+    val = string_to_trait(elem)
+    if val is None:
+        return ("None", "unknown")
+    if isinstance(val, u.UnitBase):
```
Comment (keskitalo) [inline src/toast/config/yaml.py:127]: Times like this you wish Python had a case statement

**Review by keskitalo — APPROVED** [review]
You have been busy! Looks good.

File: src/toast/config/toml.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    if isinstance(elem, int):
+        return (trait_to_string(elem), "int")
+    if isinstance(elem, float):
+        return (trait_to_string(elem), "float")
+
+    # In old TOML files, lists were stored as TOMLKit arrays.
+    # Here we cast those cases into strings to be parsed like
+    # modern containers (which are stored as strings in the
+    # the TOML files).  Remove this code after a suitable
+    # deprecation period.
```
[the remark this answers is not in the document]
Comment (tskisner) [reply src/toast/config/toml.py:40]: Done.

File: src/toast/config/utils.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def merge_config(loaded, original):
+    log = Logger.get()
+    for section, objs in loaded.items():
+        if section in original.keys():
+            # We have this section
+            if isinstance(original[section], str):
+                # Just some metadata
+                original[section] = objs
```
[the remark this answers is not in the document]
Comment (tskisner) [reply src/toast/config/utils.py:17]: In this case it is ok.  This code is just updating string objects in the top-level of the config dictionary.  The default starting config has no scalars in the top level, so this is just loaded config files overriding any top-level arbitrary strings that a previous config has set.  Those values are just user information and would not be used anywhere in the job.

File: src/toast/config/utils.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                    # Only update the value, while preserving
+                    # any pre-existing type information.
+                    for k, v in objprops.items():
+                        if k == "class":
+                            continue
+                        if k in original[section][objname]:
+                            # This key exists in the original object traits
+                            cursor = original[section][objname][k]
+                            cursor["value"] = v["value"]
+                            if "type" not in cursor:
```
[the remark this answers is not in the document]
Comment (tskisner) [reply src/toast/config/utils.py:33]: In early code, type information was often unknown in the loaded configs, but this has changed.  I updated the code to check for consistent types if they exist.

**Pull request merged**